Globals are shadowed by block parameters whose name they are a prefix of

```typescript
import Component from '@glint/environment-ember-loose/glimmer-component';

export default class Foo extends Component {
  public get bars(): Array<{ text: string }> | undefined {
    return undefined;
  }
}

import Helper from '@glint/environment-ember-loose/ember-component/helper';

declare class OrHelper<T> extends Helper<{
  readonly PositionalArgs: T[];
  readonly Return: T;
}> {}

declare module '@glint/environment-ember-loose/registry' {
  export default interface Registry {
    or: typeof OrHelper;
  }
}
```

```handlebars
{{#let (get this.bars 1) as |original|}}
  {{or original.text ""}}
{{/let}}
```

Leads to:

```
app/components/foo.hbs:2:5 - error TS2304: Cannot find name 'or'.

2   {{or original.text ""}}
      ~~
```

The crazy thing is that changing the name of the `original` variable to something else that doesn't start with `original` (say `bar`) - the error goes away. :smile:

--- a/packages/config/__tests__/environment.test.ts
+++ b/packages/config/__tests__/environment.test.ts
@@ -194,28 +194,28 @@ describe('Environments', () => {
       expect(env.getConfiguredTemplateTags()).toEqual({ internal: {} });
     });
 
+    function createEnvironment(config: string): string {
+      let name = Math.random().toString(36).slice(2);
+      let dir = `${testDir}/node_modules/@glint/environment-${name}`;
+
+      fs.mkdirSync(dir, { recursive: true });
+      fs.writeFileSync(`${dir}/env.js`, `module.exports = ${config}`);
+      fs.writeFileSync(
+        `${dir}/package.json`,
+        JSON.stringify({
+          name: `@glint/environment-${name}`,
+          'glint-environment': 'env',
+        })
+      );
+
+      return name;
+    }
+
     describe('merging multiple environments', () => {
-      function createEnvironment(config: string): string {
-        let name = Math.random().toString(36).slice(2);
-        let dir = `${testDir}/node_modules/@glint/environment-${name}`;
-
-        fs.mkdirSync(dir, { recursive: true });
-        fs.writeFileSync(`${dir}/env.js`, `module.exports = () => (${config})`);
-        fs.writeFileSync(
-          `${dir}/package.json`,
-          JSON.stringify({
-            name: '@glint/environment-a',
-            'glint-environment': 'env',
-          })
-        );
-
-        return name;
-      }
-
       test('loading compatible environments', () => {
-        let envA = createEnvironment('{ tags: { "foo-bar": { hbs: {} } } }');
+        let envA = createEnvironment('() => ({ tags: { "foo-bar": { hbs: {} } } })');
         let envB = createEnvironment(
-          '{ tags: { "foo-bar": { tpl: {} }, "baz": { hbs: {} } }, template: { typesPath: "foo" } }'
+          '() => ({ tags: { "foo-bar": { tpl: {} }, "baz": { hbs: {} } }, template: { typesPath: "foo" } })'
         );
 
         let env = GlintEnvironment.load([envA, envB], { rootDir: testDir });
@@ -228,8 +228,8 @@ describe('Environments', () => {
       });
 
       test('loading conflicting standalone template config', () => {
-        let envA = createEnvironment('{ template: { typesPath: "foo" } }');
-        let envB = createEnvironment('{ template: { typesPath: "bar" } }');
+        let envA = createEnvironment('() => ({ template: { typesPath: "foo" } })');
+        let envB = createEnvironment('() => ({ template: { typesPath: "bar" } })');
 
         expect(() => GlintEnvironment.load([envA, envB], { rootDir: testDir })).toThrow(
           'Multiple configured Glint environments attempted to define behavior for standalone template files'
@@ -237,12 +237,43 @@ describe('Environments', () => {
       });
 
       test('loading conflicting tags config', () => {
-        let envA = createEnvironment('{ tags: { foo: { hbs: {} } } }');
-        let envB = createEnvironment('{ tags: { foo: { hbs: {} } } }');
+        let envA = createEnvironment('() => ({ tags: { foo: { hbs: {} } } })');
+        let envB = createEnvironment('() => ({ tags: { foo: { hbs: {} } } })');
 
         expect(() => GlintEnvironment.load([envA, envB], { rootDir: testDir })).toThrow(
           "Multiple configured Glint environments attempted to define behavior for the tag `hbs` in module 'foo'"
         );
+      });
+    });
+
+    describe('environment user configuration', () => {
+      test('single string', () => {
+        let envName = createEnvironment('cfg => ({ tags: { foo: { hbs: cfg } } })');
+        let env = GlintEnvironment.load(envName, { rootDir: testDir });
+
+        expect(env.getConfiguredTemplateTags()).toEqual({ foo: { hbs: {} } });
+      });
+
+      test('string array', () => {
+        let aName = createEnvironment('cfg => ({ tags: { foo: { hbs: cfg } } })');
+        let bName = createEnvironment('cfg => ({ tags: { bar: { hbs: cfg } } })');
+        let env = GlintEnvironment.load([aName, bName], { rootDir: testDir });
+
+        expect(env.getConfiguredTemplateTags()).toEqual({ foo: { hbs: {} }, bar: { hbs: {} } });
+      });
+
+      test('object with no explicit config', () => {
+        let envName = createEnvironment('cfg => ({ tags: { foo: { hbs: cfg } } })');
+        let env = GlintEnvironment.load({ [envName]: {} }, { rootDir: testDir });
+
+        expect(env.getConfiguredTemplateTags()).toEqual({ foo: { hbs: {} } });
+      });
+
+      test('object with explicit config', () => {
+        let envName = createEnvironment('cfg => ({ tags: { foo: { hbs: cfg } } })');
+        let env = GlintEnvironment.load({ [envName]: { custom: true } }, { rootDir: testDir });
+
+        expect(env.getConfiguredTemplateTags()).toEqual({ foo: { hbs: { custom: true } } });
       });
     });
   });

--- a/packages/config/__tests__/environment.test.ts
+++ b/packages/config/__tests__/environment.test.ts
@@ -7,7 +7,7 @@ describe('Environments', () => {
     test('locating a single tag', () => {
       let env = new GlintEnvironment(['test-env'], {
         tags: {
-          'my-cool-environment': { hbs: { capturesOuterScope: false, typesSource: 'whatever' } },
+          'my-cool-environment': { hbs: { typesSource: 'whatever' } },
         },
       });
 
@@ -19,9 +19,9 @@ describe('Environments', () => {
     test('locating one of several tags', () => {
       let env = new GlintEnvironment(['test-env'], {
         tags: {
-          'my-cool-environment': { hbs: { capturesOuterScope: false, typesSource: 'whatever' } },
-          'another-env': { tagMe: { capturesOuterScope: false, typesSource: 'over-here' } },
-          'and-this-one': { hbs: { capturesOuterScope: false, typesSource: '✨' } },
+          'my-cool-environment': { hbs: { typesSource: 'whatever' } },
+          'another-env': { tagMe: { typesSource: 'over-here' } },
+          'and-this-one': { hbs: { typesSource: '✨' } },
         },
       });
 
@@ -33,7 +33,7 @@ describe('Environments', () => {
     test('checking a module with no tags in use', () => {
       let env = new GlintEnvironment(['test-env'], {
         tags: {
-          'my-cool-environment': { hbs: { capturesOuterScope: false, typesSource: 'whatever' } },
+          'my-cool-environment': { hbs: { typesSource: 'whatever' } },
         },
       });
 
@@ -46,7 +46,6 @@ describe('Environments', () => {
       let tags = {
         '@glimmerx/component': {
           hbs: {
-            capturesOuterScope: false,
             typesSource: '@glint/environment-glimmerx/-private/dsl',
           },
         },

--- a/packages/config/src/config.ts
+++ b/packages/config/src/config.ts
@@ -3,7 +3,7 @@ import { Minimatch, IMinimatch } from 'minimatch';
 import { GlintEnvironment } from './environment';
 
 export type GlintConfigInput = {
-  environment: string | Array<string>;
+  environment: string | Array<string> | Record<string, unknown>;
   checkStandaloneTemplates?: boolean;
   include?: string | Array<string>;
   exclude?: string | Array<string>;
@@ -90,8 +90,10 @@ function validateConfigInput(input: Record<string, unknown>): asserts input is G
   assert(
     Array.isArray(input['environment'])
       ? input['environment'].every((env) => typeof env === 'string')
-      : typeof input['environment'] === 'string',
-    'Glint config must specify an `environment` that is a string or an array of strings'
+      : typeof input['environment'] === 'string' ||
+          (typeof input['environment'] === 'object' && input['environment']),
+    'Glint config must specify an `environment` that is a string, array of strings, or an object ' +
+      'mapping environment names to their config.'
   );
 
   assert(

--- a/packages/config/src/environment.ts
+++ b/packages/config/src/environment.ts
@@ -70,10 +70,13 @@ export class GlintEnvironment {
     this.templateExtensions = this.extensionsOfType('template');
   }
 
-  public static load(name: string | Array<string>, { rootDir = '.' } = {}): GlintEnvironment {
-    let names = Array.isArray(name) ? name : [name];
-    let config = loadMergedEnvironmentConfig(names, rootDir);
-    return new GlintEnvironment(names, config);
+  public static load(
+    specifier: string | Array<string> | Record<string, unknown>,
+    { rootDir = '.' } = {}
+  ): GlintEnvironment {
+    let envs = normalizeEnvironmentSpecifier(specifier);
+    let config = loadMergedEnvironmentConfig(envs, rootDir);
+    return new GlintEnvironment(Object.keys(envs), config);
   }
 
   public getSourceKind(fileName: string): SourceKind | 'unknown' {
@@ -175,17 +178,29 @@ export class GlintEnvironment {
   }
 }
 
+function normalizeEnvironmentSpecifier(
+  specifier: string | string[] | Record<string, unknown>
+): Record<string, unknown> {
+  if (typeof specifier === 'string') {
+    return { [specifier]: null };
+  } else if (Array.isArray(specifier)) {
+    return specifier.reduce((obj, name) => ({ ...obj, [name]: null }), {});
+  }
+
+  return specifier;
+}
+
 function loadMergedEnvironmentConfig(
-  envNames: Array<string>,
+  envs: Record<string, unknown>,
   rootDir: string
 ): GlintEnvironmentConfig {
   let tags: GlintTagsConfig = {};
   let extensions: GlintExtensionsConfig = { ...DEFAULT_EXTENSIONS };
   let template: GlintTemplateConfig | undefined;
-  for (let name of envNames) {
-    let envModule = require(locateEnvironment(name, rootDir));
+  for (let [envName, envUserConfig] of Object.entries(envs)) {
+    let envModule = require(locateEnvironment(envName, rootDir));
     let envFunction = envModule.default ?? envModule;
-    let config = envFunction() as GlintEnvironmentConfig;
+    let config = envFunction(envUserConfig ?? {}) as GlintEnvironmentConfig;
 
     if (config.template) {
       if (template) {

--- a/packages/config/src/environment.ts
+++ b/packages/config/src/environment.ts
@@ -25,7 +25,7 @@ export type GlintExtensionsConfig = {
 
 export type GlintTagConfig = {
   typesSource: string;
-  capturesOuterScope: boolean;
+  globals?: Array<string>;
 };
 
 export type GlintTagsConfig = {

--- a/packages/core/__tests__/language-server/completions.test.ts
+++ b/packages/core/__tests__/language-server/completions.test.ts
@@ -16,21 +16,21 @@ describe('Language Server: Completions', () => {
 
   test('querying a standalone template', () => {
     project.write('.glintrc', 'environment: ember-loose');
-    project.write('index.hbs', '<Foo as |foo|>{{fo}}</Foo>');
+    project.write('index.hbs', '<LinkT />');
 
     let server = project.startLanguageServer();
     let completions = server.getCompletions(project.fileURI('index.hbs'), {
       line: 0,
-      character: 17,
+      character: 6,
     });
 
-    let completion = completions?.find((item) => item.label === 'foo');
+    let completion = completions?.find((item) => item.label === 'LinkTo');
 
-    expect(completion?.kind).toEqual(CompletionItemKind.Variable);
+    expect(completion?.kind).toEqual(CompletionItemKind.Field);
 
     let details = server.getCompletionDetails(completion!);
 
-    expect(details.detail).toEqual('const foo: any');
+    expect(details.detail).toEqual('(property) Globals.LinkTo: LinkToComponent');
   });
 
   test('passing component args', () => {
@@ -151,33 +151,6 @@ describe('Language Server: Completions', () => {
     let details = server.getCompletionDetails(letterCompletion!);
 
     expect(details.detail).toEqual('const letter: string');
-  });
-
-  test('globals', () => {
-    let code = stripIndent`
-      import Component, { hbs } from '@glint/environment-glimmerx/component';
-
-      export default class MyComponent extends Component {
-        static template = hbs\`
-          {{deb}}
-        \`;
-      }
-    `;
-
-    project.write('index.ts', code);
-
-    let server = project.startLanguageServer();
-    let completions = server.getCompletions(project.fileURI('index.ts'), {
-      line: 4,
-      character: 9,
-    });
-
-    let completion = completions?.find((completion) => completion.label === 'debugger');
-
-    expect(completion).toMatchObject({
-      kind: CompletionItemKind.Field,
-      label: 'debugger',
-    });
   });
 
   test('referencing module-scope identifiers', async () => {

--- a/packages/environment-ember-loose/-private/environment/index.ts
+++ b/packages/environment-ember-loose/-private/environment/index.ts
@@ -17,7 +17,6 @@ export default function glimmerxEnvironment(): GlintEnvironmentConfig {
       'ember-cli-htmlbars': {
         hbs: {
           typesSource: '@glint/environment-ember-loose/-private/dsl',
-          capturesOuterScope: false,
         },
       },
     },

--- a/packages/environment-glimmerx/-private/environment/index.ts
+++ b/packages/environment-glimmerx/-private/environment/index.ts
@@ -1,20 +1,33 @@
-import { GlintEnvironmentConfig } from '@glint/config';
+import { GlintEnvironmentConfig, GlintTagConfig } from '@glint/config';
 
-export default function glimmerxEnvironment(): GlintEnvironmentConfig {
+export default function glimmerxEnvironment(
+  config: Record<string, unknown>
+): GlintEnvironmentConfig {
+  let additionalGlobals = Array.isArray(config['additionalGlobals'])
+    ? config['additionalGlobals']
+    : [];
+
+  let tagConfig: GlintTagConfig = {
+    typesSource: '@glint/environment-glimmerx/-private/dsl',
+    globals: [
+      'debugger',
+      'each',
+      'has-block',
+      'has-block-params',
+      'if',
+      'in-element',
+      'let',
+      'unless',
+      'with',
+      'yield',
+      ...additionalGlobals,
+    ],
+  };
+
   return {
     tags: {
-      '@glimmerx/component': {
-        hbs: {
-          typesSource: '@glint/environment-glimmerx/-private/dsl',
-          capturesOuterScope: true,
-        },
-      },
-      '@glint/environment-glimmerx/component': {
-        hbs: {
-          typesSource: '@glint/environment-glimmerx/-private/dsl',
-          capturesOuterScope: true,
-        },
-      },
+      '@glimmerx/component': { hbs: tagConfig },
+      '@glint/environment-glimmerx/component': { hbs: tagConfig },
     },
   };
 }

--- a/packages/transform/__tests__/offset-mapping.test.ts
+++ b/packages/transform/__tests__/offset-mapping.test.ts
@@ -14,18 +14,11 @@ describe('Source-to-source offset mapping', () => {
     transformedModule: TransformedModule;
   };
 
-  function rewriteInlineTemplate({
-    contents,
-    identifiersInScope = [],
-  }: {
-    contents: string;
-    identifiersInScope?: string[];
-  }): RewrittenTestModule {
+  function rewriteInlineTemplate({ contents }: { contents: string }): RewrittenTestModule {
     let script = {
       filename: 'test.ts',
       contents: stripIndent`
         import Component, { hbs } from '@glint/environment-glimmerx/component';
-        import { ${identifiersInScope.join(', ')} } from 'dummy';
 
         export default class MyComponent extends Component {
           static template = hbs\`
@@ -180,7 +173,6 @@ describe('Source-to-source offset mapping', () => {
     describe('path segments', () => {
       test('simple in-scope paths', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['foo'],
           contents: '{{foo.bar}}',
         });
         expectTokenMapping(module, 'foo');
@@ -217,7 +209,6 @@ describe('Source-to-source offset mapping', () => {
     describe('keys', () => {
       test('named params to mustaches', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['foo', 'hello'],
           contents: '{{foo bar=hello}}',
         });
         expectTokenMapping(module, 'foo');
@@ -227,7 +218,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('named spinal-case params to mustaches', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['foo', 'hello'],
           contents: '{{foo bar-baz=hello}}',
         });
         expectTokenMapping(module, 'foo');
@@ -237,7 +227,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('component args', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['Foo', 'hello'],
           contents: '<Foo @bar={{hello}} />',
         });
         expectTokenMapping(module, 'Foo');
@@ -247,7 +236,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('spinal-case component args', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['Foo', 'hello'],
           contents: '<Foo @bar-baz={{hello}} />',
         });
         expectTokenMapping(module, 'Foo');
@@ -257,7 +245,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('named blocks', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['Foo'],
           contents: '<Foo><:blockName>hi</:blockName></Foo>',
         });
         expectTokenMapping(module, 'Foo');
@@ -266,7 +253,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('spinal-case named blocks', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['Foo'],
           contents: '<Foo><:block-name>hi</:block-name></Foo>',
         });
         expectTokenMapping(module, 'Foo');
@@ -294,7 +280,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('angle bracket params', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['Foo'],
           contents: '<Foo as |bar baz|></Foo>',
         });
         expectTokenMapping(module, 'Foo');
@@ -306,7 +291,6 @@ describe('Source-to-source offset mapping', () => {
     describe('block mustaches', () => {
       test('simple identifiers', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['foo'],
           contents: '{{#foo}}{{/foo}}',
         });
         expectTokenMapping(module, 'foo', { occurrence: 0 });
@@ -315,7 +299,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('simple paths', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['foo'],
           contents: '{{#foo.bar}}{{/foo.bar}}',
         });
         expectTokenMapping(module, 'foo', { occurrence: 0 });
@@ -344,7 +327,6 @@ describe('Source-to-source offset mapping', () => {
     describe('angle bracket components', () => {
       test('simple identifiers', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['Foo'],
           contents: '<Foo></Foo>',
         });
         expectTokenMapping(module, 'Foo', { occurrence: 0 });
@@ -353,7 +335,6 @@ describe('Source-to-source offset mapping', () => {
 
       test('simple paths', () => {
         let module = rewriteInlineTemplate({
-          identifiersInScope: ['foo'],
           contents: '<foo.bar></foo.bar>',
         });
         expectTokenMapping(module, 'foo', { occurrence: 0 });
@@ -490,8 +471,8 @@ describe('Diagnostic offset mapping', () => {
       code,
       messageText,
       file: transformedContentsFile,
-      start: transformedModule.transformedContents.indexOf('"foo"'),
-      length: 5,
+      start: transformedModule.transformedContents.indexOf('foo'),
+      length: 3,
     };
 
     let rewritten = rewriteDiagnostic(ts, original, () => transformedModule);

--- a/packages/transform/__tests__/rewrite.test.ts
+++ b/packages/transform/__tests__/rewrite.test.ts
@@ -117,8 +117,9 @@ describe('rewriteModule', () => {
       let testEnvironment = new GlintEnvironment(['test'], {
         tags: {
           '@glint/test-env': {
-            hbsCapture: { typesSource: '@glint/test-env', capturesOuterScope: true },
-            hbsIgnore: { typesSource: '@glint/test-env', capturesOuterScope: false },
+            hbsCaptureAll: { typesSource: '@glint/test-env', globals: [] },
+            hbsCaptureSome: { typesSource: '@glint/test-env', globals: ['global'] },
+            hbsCaptureNone: { typesSource: '@glint/test-env' },
           },
         },
       });
@@ -126,12 +127,13 @@ describe('rewriteModule', () => {
       let script = {
         filename: 'test.ts',
         contents: stripIndent`
-          import { hbsCapture, hbsIgnore } from '@glint/test-env';
+          import { hbsCaptureAll, hbsCaptureSome, hbsCaptureNone } from '@glint/test-env';
 
           const message = 'hello';
 
-          hbsCapture\`{{message}}\`;
-          hbsIgnore\`{{message}}\`;
+          hbsCaptureAll\`{{global}} {{message}}\`;
+          hbsCaptureSome\`{{global}} {{message}}\`;
+          hbsCaptureNone\`{{global}} {{message}}\`;
         `,
       };
 
@@ -139,17 +141,25 @@ describe('rewriteModule', () => {
 
       expect(transformedModule?.errors).toEqual([]);
       expect(transformedModule?.transformedContents).toMatchInlineSnapshot(`
-        "import { hbsCapture, hbsIgnore } from '@glint/test-env';
+        "import { hbsCaptureAll, hbsCaptureSome, hbsCaptureNone } from '@glint/test-env';
 
         const message = 'hello';
 
         ({} as typeof import(\\"@glint/test-env\\")).template(function(𝚪, χ: typeof import(\\"@glint/test-env\\")) {
-          hbsCapture;
+          hbsCaptureAll;
+          χ.emitValue(χ.resolveOrReturn(global)({}));
           χ.emitValue(χ.resolveOrReturn(message)({}));
           𝚪; χ;
         });
         ({} as typeof import(\\"@glint/test-env\\")).template(function(𝚪, χ: typeof import(\\"@glint/test-env\\")) {
-          hbsIgnore;
+          hbsCaptureSome;
+          χ.emitValue(χ.resolveOrReturn(χ.Globals[\\"global\\"])({}));
+          χ.emitValue(χ.resolveOrReturn(message)({}));
+          𝚪; χ;
+        });
+        ({} as typeof import(\\"@glint/test-env\\")).template(function(𝚪, χ: typeof import(\\"@glint/test-env\\")) {
+          hbsCaptureNone;
+          χ.emitValue(χ.resolveOrReturn(χ.Globals[\\"global\\"])({}));
           χ.emitValue(χ.resolveOrReturn(χ.Globals[\\"message\\"])({}));
           𝚪; χ;
         });"

--- a/packages/transform/__tests__/template-to-typescript.test.ts
+++ b/packages/transform/__tests__/template-to-typescript.test.ts
@@ -390,7 +390,7 @@ describe('rewriteTemplate', () => {
           {{log (array 1 true "free")}}
         `;
 
-        expect(templateBody(template, { identifiersInScope: ['log'] })).toMatchInlineSnapshot(
+        expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(
           `"χ.emitValue(χ.resolve(log)({}, [1, true, \\"free\\"]));"`
         );
       });
@@ -423,7 +423,7 @@ describe('rewriteTemplate', () => {
           {{log (hash a=1 b="ok")}}
         `;
 
-        expect(templateBody(template, { identifiersInScope: ['log'] })).toMatchInlineSnapshot(`
+        expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
           "χ.emitValue(χ.resolve(log)({}, ({
             a: 1,
             b: \\"ok\\",
@@ -445,28 +445,25 @@ describe('rewriteTemplate', () => {
         });
 
         test('in-scope identifier', () => {
-          let identifiersInScope = ['message'];
           let template = '{{message}}';
 
-          expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(
+          expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(
             `"χ.emitValue(χ.resolveOrReturn(message)({}));"`
           );
         });
 
         test('chained path', () => {
-          let identifiersInScope = ['obj'];
           let template = '{{obj.foo.bar}}';
 
-          expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(
+          expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(
             `"χ.emitValue(χ.resolveOrReturn(obj?.foo?.bar)({}));"`
           );
         });
 
         test('chained path with a spinal-case key', () => {
-          let identifiersInScope = ['obj'];
           let template = '{{obj.foo-bar.baz}}';
 
-          expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(
+          expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(
             `"χ.emitValue(χ.resolveOrReturn(obj?.[\\"foo-bar\\"]?.baz)({}));"`
           );
         });
@@ -496,10 +493,9 @@ describe('rewriteTemplate', () => {
         });
 
         test('passed as an attr', () => {
-          let identifiersInScope = ['Foo', 'helper'];
           let template = '<Foo data-bar={{helper param=true}} />';
 
-          expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+          expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
             "{
               const 𝛄 = χ.emitComponent(χ.resolve(Foo)({}));
               χ.applyAttributes(𝛄.element, {
@@ -510,10 +506,9 @@ describe('rewriteTemplate', () => {
         });
 
         test('passed as an @arg', () => {
-          let identifiersInScope = ['Foo', 'helper'];
           let template = '<Foo @bar={{helper param=true}} />';
 
-          expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+          expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
             "{
               const 𝛄 = χ.emitComponent(χ.resolve(Foo)({ bar: χ.resolve(helper)({ param: true }) }));
               𝛄;
@@ -558,10 +553,9 @@ describe('rewriteTemplate', () => {
         });
 
         test('as an @arg value', () => {
-          let identifiersInScope = ['Greet'];
           let template = '<Greet @message={{@arg}} />';
 
-          expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+          expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
             "{
               const 𝛄 = χ.emitComponent(χ.resolve(Greet)({ message: 𝚪.args.arg }));
               𝛄;
@@ -607,28 +601,25 @@ describe('rewriteTemplate', () => {
 
     describe('helper and inline-curly component invocations', () => {
       test('positional params', () => {
-        let identifiersInScope = ['doSomething'];
         let template = '{{doSomething "hello" 123}}';
 
-        expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(
+        expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(
           `"χ.emitValue(χ.resolve(doSomething)({}, \\"hello\\", 123));"`
         );
       });
 
       test('named params', () => {
-        let identifiersInScope = ['doSomething'];
         let template = '{{doSomething a=123 b="ok"}}';
 
-        expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(
+        expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(
           `"χ.emitValue(χ.resolve(doSomething)({ a: 123, b: \\"ok\\" }));"`
         );
       });
 
       test('named and positional params', () => {
-        let identifiersInScope = ['doSomething'];
         let template = '{{doSomething "one" true 3 four=4}}';
 
-        expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(
+        expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(
           `"χ.emitValue(χ.resolve(doSomething)({ four: 4 }, \\"one\\", true, 3));"`
         );
       });
@@ -637,10 +628,9 @@ describe('rewriteTemplate', () => {
 
   describe('modifiers', () => {
     test('on a plain element', () => {
-      let identifiersInScope = ['modifier'];
       let template = `<div {{modifier foo="bar"}}></div>`;
 
-      expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+      expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
         "{
           const 𝛄 = χ.emitElement(\\"div\\");
           χ.applyModifier(𝛄.element, χ.resolve(modifier)({ foo: \\"bar\\" }));
@@ -649,10 +639,9 @@ describe('rewriteTemplate', () => {
     });
 
     test('on a component', () => {
-      let identifiersInScope = ['MyComponent', 'modifier'];
       let template = `<MyComponent {{modifier foo="bar"}}/>`;
 
-      expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+      expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
         "{
           const 𝛄 = χ.emitComponent(χ.resolve(MyComponent)({}));
           χ.applyModifier(𝛄.element, χ.resolve(modifier)({ foo: \\"bar\\" }));
@@ -663,10 +652,9 @@ describe('rewriteTemplate', () => {
 
   describe('subexpressions', () => {
     test('resolution', () => {
-      let identifiersInScope = ['concat', 'foo'];
       let template = `<div data-attr={{concat (foo 1) (foo true)}}></div>`;
 
-      expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+      expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
         "{
           const 𝛄 = χ.emitElement(\\"div\\");
           χ.applyAttributes(𝛄.element, {
@@ -835,10 +823,9 @@ describe('rewriteTemplate', () => {
     });
 
     test('with splattributes', () => {
-      let identifiersInScope = ['Foo'];
       let template = '<Foo ...attributes />';
 
-      expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+      expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
         "{
           const 𝛄 = χ.emitComponent(χ.resolve(Foo)({}));
           χ.applySplattributes(𝚪.element, 𝛄.element);
@@ -847,10 +834,9 @@ describe('rewriteTemplate', () => {
     });
 
     test('with a path for a name', () => {
-      let identifiersInScope = ['foo'];
       let template = '<foo.bar @arg="hello" />';
 
-      expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+      expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
         "{
           const 𝛄 = χ.emitComponent(χ.resolve(foo?.bar)({ arg: \\"hello\\" }));
           𝛄;
@@ -918,10 +904,9 @@ describe('rewriteTemplate', () => {
     });
 
     test('with concat args', () => {
-      let identifiersInScope = ['Foo', 'baz'];
       let template = `<Foo @arg="bar-{{baz}}" />`;
 
-      expect(templateBody(template, { identifiersInScope })).toMatchInlineSnapshot(`
+      expect(templateBody(template, { globals: [] })).toMatchInlineSnapshot(`
         "{
           const 𝛄 = χ.emitComponent(χ.resolve(Foo)({ arg: \`\${χ.emitValue(χ.resolveOrReturn(baz)({}))}\` }));
           𝛄;

--- a/packages/transform/src/inlining/tagged-strings.ts
+++ b/packages/transform/src/inlining/tagged-strings.ts
@@ -21,7 +21,7 @@ export function calculateTaggedTemplateSpans(
 
   let tagConfig = getConfigForTag(tag, environment);
   if (tagConfig) {
-    let { typesSource, capturesOuterScope } = tagConfig;
+    let { typesSource, globals } = tagConfig;
     let tagName = tag.node.name;
     let { quasis } = path.node.quasi;
 
@@ -44,11 +44,10 @@ export function calculateTaggedTemplateSpans(
     let preamble = [`${tagName};`];
 
     let { inClass, className, typeParams, contextType } = getContainingTypeInfo(path);
-    let identifiersInScope = capturesOuterScope ? Object.keys(path.scope.getAllBindings()) : [];
     let transformedTemplate = templateToTypescript(template, {
       typesPath: typesSource,
       preamble,
-      identifiersInScope,
+      globals,
       typeParams,
       contextType,
       useJsDoc: environment.isUntypedScript(script.filename),

--- a/packages/transform/src/scope-stack.ts
+++ b/packages/transform/src/scope-stack.ts
@@ -28,20 +28,6 @@ export default class ScopeStack {
     return this.top.has(identifier);
   }
 
-  public hasMatchingBinding(prefix: string): boolean {
-    if (this.hasBinding(prefix)) {
-      return true;
-    }
-
-    for (let identifier of this.top) {
-      if (identifier.startsWith(prefix)) {
-        return true;
-      }
-    }
-
-    return false;
-  }
-
   private get top(): Set<string> {
     return this.stack[0];
   }

--- a/packages/transform/src/template-to-typescript.ts
+++ b/packages/transform/src/template-to-typescript.ts
@@ -12,7 +12,7 @@ type BlockKeyword = typeof BLOCK_KEYWORDS[number];
 
 export type TemplateToTypescriptOptions = {
   typesPath: string;
-  identifiersInScope?: Array<string>;
+  globals?: Array<string> | undefined;
   contextType?: string;
   typeParams?: string;
   preamble?: Array<string>;
@@ -28,7 +28,7 @@ export function templateToTypescript(
   template: string,
   {
     typesPath,
-    identifiersInScope = [],
+    globals,
     typeParams = '',
     contextType,
     preamble = [],
@@ -36,7 +36,7 @@ export function templateToTypescript(
   }: TemplateToTypescriptOptions
 ): RewriteResult {
   return mapTemplateContents(template, (ast, { emit, record, rangeForLine, rangeForNode }) => {
-    let scope = new ScopeStack(identifiersInScope);
+    let scope = new ScopeStack([]);
 
     emitTemplateBoilerplate(useJsDoc, () => {
       for (let line of preamble) {
@@ -347,12 +347,27 @@ export function templateToTypescript(
     }
 
     function emitIdentifierReference(name: string, hbsOffset: number): void {
-      if (scope.hasMatchingBinding(name)) {
-        emit.identifier(name, hbsOffset);
-      } else {
+      if (treatAsGlobal(name)) {
         emit.text('χ.Globals["');
         emit.identifier(JSON.stringify(name).slice(1, -1), hbsOffset, name.length);
         emit.text('"]');
+      } else {
+        emit.identifier(name, hbsOffset);
+      }
+    }
+
+    function treatAsGlobal(name: string): boolean {
+      if (globals) {
+        // If we have a known set of global identifiers, we should only treat
+        // members of that set as global and assume everything else is local.
+        // This is typically true in environments that capture scope, like
+        // GlimmerX or strict-mode Ember.
+        return globals.includes(name);
+      } else {
+        // Otherwise, we assume everything is global unless we can see it
+        // in scope as a block variable. This is the case in resolver-based
+        // environments like loose-mode Ember.
+        return !scope.hasBinding(name);
       }
     }
 

